### PR TITLE
Add option to ADS to return hidden workspaces in getObjects()

### LINKED ISSUE #22901: Project Recovery throws unexpected exceptions
**Original reporter:** Peter Baker/Anthony Lim

### Expected behavior

If project recovery fails to reconstruct the history, it should simply throw an error saying that recovery was not possible.

### Actual behavior

Failure of project recovery results in an unexpected error

### Steps to reproduce the behavior

* Make sure the recovery directory is empty and Mantid is not running
* Unzip the file below to the project recovery directoy
* Start Mantid
* When offered recovery choose yes
* Mantid throws an unexpected error
[2018-07-12T16-33-16.zip](https://github.com/mantidproject/mantid/files/2189268/2018-07-12T16-33-16.zip)

### PR DESCRIPTION
**Pull request by DavidFair**
**Description of work.**
Adds a parameter to the ADS `getObjects` method similar to that found in `getObjectNames`. 

This allows users to specify whether to include hidden workspaces or retain the default behaviour of using the config option.

**To test:**
- Check unit tests pass
- Create a sample workspace with a hidden name (for example `__foo`)
- Ensure a project recovery checkpoint contains this after saving

Fixes #22901. 

Does this update require release notes?
- [ ] Yes
- [x] No

---

#### Reviewer ####

Please comment on the following ([full description](http://www.mantidproject.org/Individual_Ticket_Testing)):

##### Code Review #####

- [ ] Is the code of an acceptable quality?
- [ ] Does the code conform to the [coding standards](http://www.mantidproject.org/Coding_Standards)?
- [ ] Are the unit tests small and test the class in isolation?
- [ ] If there are changes in the release notes then do they describe the changes appropriately?

##### Functional Tests #####

- [ ] Do changes function as described? Add comments below that describe the tests performed?
- [ ] Do the changes handle unexpected situations, e.g. bad input?
- [ ] Has the relevant (user and developer) documentation been added/updated?

Does everything look good? Mark the review as **Approve**. A member of `@mantidproject/gatekeepers` will take care of it.

### CODE REVIEW

File: Framework/Kernel/inc/MantidKernel/DataService.h [hunk 1]
```diff
@@ -482,15 +482,26 @@ template <typename T> class DLLExport DataService {
...
 
   /// Get a vector of the pointers to the data objects stored by the service
-  std::vector<boost::shared_ptr<T>> getObjects() const {
+  std::vector<boost::shared_ptr<T>>
+  getObjects(DataServiceHidden includeHidden = DataServiceHidden::Auto) const {
     std::lock_guard<std::recursive_mutex> _lock(m_mutex);
 
-    const bool showingHidden = showingHiddenObjects();
+    if (includeHidden == DataServiceHidden::Include) {
+      // Return all elements including hidden ones
```
Comment (Mantid-Matthew) [inline Framework/Kernel/inc/MantidKernel/DataService.h]: Does this if block do anything currently ?

Comment (DavidFair) [reply Framework/Kernel/inc/MantidKernel/DataService.h]: No, good catch - it's a leftover from before refactoring

**Comment by Mantid-Matthew** [comment]
The following work-flow currently does not work for me.
* Load a workspace.
* Rename the workspace to start with `__` i.e. make it hidden
* Segfault Mantid

I found that either no project recovery was prompted when I restarted mantid or the session was restored to after the loading but before the renaming. Also of note was the fact that when looking at the debug log project recovery claimed there was nothing to save

**Comment by Mantid-Matthew** [comment]
This appears to occur when Mantid contains only hidden workspaces. Upon discussion it was decided that this was the desired behavior as it is confusing if project recovery runs to no visible result.

**Review by Matthew-Andrew — APPROVED** [review]
All looks good to me.

**Comment by Mantid-Matthew** [comment]
@DavidFair this seems to be failing MantidPlotProjectSerialiseTest on windows which is close enough to what you are working on that I'm wary of dismissing it as en erroneous failure. Any ideas ?

File: Framework/Kernel/test/DataServiceTest.h [hunk 2]
```diff
@@ -6,6 +6,9 @@
 #include <cxxtest/TestSuite.h>
 #include <Poco/NObserver.h>
 #include <boost/make_shared.hpp>
+
+#include <iostream>
```
Comment (martyngigg) [inline Framework/Kernel/test/DataServiceTest.h]: Can we remove this header? It's a big header and can easily get left around when it's not required.

**Comment by Mantid-Matthew** [comment]
The other test failures are just more ICat failures.

**Comment by DavidFair** [comment]
@Mantid-Matthew - This was a spurious failure `IOError: Could not clean up folder after test`

**Comment by NickDraper** [comment]
ICAT failures only due to the system being down.

**Pull request merged**